api/agent/drivers/docker/docker.go

```diff
@@ -51,17 +51,79 @@ func (r *runResult) Error() error   { return r.err }
 func (r *runResult) Status() string { return r.status }
 
 type DockerDriver struct {
-	cancel   func()
-	conf     drivers.Config
-	docker   dockerClient // retries on *docker.Client, restricts ad hoc *docker.Client usage / retries
-	hostname string
-	auths    map[string]driverAuthConfig
-	pool     DockerPool
+	cancel     func()
+	conf       drivers.Config
+	docker     dockerClient // retries on *docker.Client, restricts ad hoc *docker.Client usage / retries
+	hostname   string
+	auths      map[string]driverAuthConfig
+	pool       DockerPool
+	imageCache *Cache
 	// protects networks map
 	networksLock sync.Mutex
 	networks     map[string]uint64
 }
 
+// NewImageCleaner builds an evicter that loops every ImageCacheCleanInterval, checkes to see
+// if there is more space consumed then the maxSize allowed by configuration. If there is
+// NewImageCleaner checks the image cache for the list of evicitable images and then tries
+// in order of most evicitable to remove the image. If the disk consumption constraint is
+// satisfied the for loop breaks.
+func NewImageCleaner(context context.Context, dockerDriver *DockerDriver, maxSize int64) error {
+	opts := docker.RemoveImageOptions{}
+	duopts := docker.DiskUsageOptions{}
+	opts.Force = true
+	opts.NoPrune = false
+	opts.Context = context
+	duopts.Context = context
+	logrus.Info("Starting image cleaner")
+	ticker := time.NewTicker(dockerDriver.conf.ImageCacheCleanInterval)
+	for {
+		select {
+		case <-context.Done():
+			return nil
+		case <-ticker.C:
+			du, err := dockerDriver.docker.DiskUsage(duopts)
+			if err != nil {
+				logrus.WithError(err).Error("attempting to check disk usage")
+			}
+			if du.LayersSize > maxSize {
+				toEvict := dockerDriver.imageCache.Evictable()
+				for _, i := range toEvict {
```

Comment:
is there a race here? We get list of evictable, but these could then be added back as not-evictable while we are attempting to remove them?

Reply:
yea, need to hold lock over removal, also need lock over all dependent parent images due to docker rm behavior, which may be tricky. eg where each letter is an image:

     a
   /   \
x      y

there's some ordering of events that goes like this: x is the only child of base image a, x is scheduled for eviction, a is now a node with no edges and docker can remove it, concurrently to that removal of x or a, lock is taken on y, which also is a child of a, a is removed while y is pulled, this results in container creation failing, which may result in user felt errors / added latency as well as possible weird states in the image eviction cache we need to handle. not clear how best to handle this, but it's the main tricky problem we need to solve here - as it is, locks are only held over function end images and not any layers it may depend on - this could be okay ultimately if we flesh out the behavior really well, but it's potentially tricky.

Reply:
another thing that comes to mind is that it's possible that an agent initiates an image removal, agent panics/dies and restarts, then restarts and loads image into image cache and it ends up getting removed even while it has the lock held in its own process. this one is pretty whack, makes me think we should try to use all docker commands to manage without having our own locks too if we can, but it needs some reasoning about... if we retry container create on certain errors we may in the worst case have a container that fails a container create and retries, it's added latency but it's possibly correct without us having to manage our own lock process over our own logical docker image cache replica we're managing... need to think on it tho